text/0000-relax-const-restrictions.md

```diff
@@ -0,0 +1,94 @@
+
+This makes the context of a function observable behavior.
+
+If a `const fn` is called in a runtime context, no additional restrictions are applied to it, and it may do anything a non-`const fn` can (for example, calling into FFI).
+
+A `const fn` being called in a const context will still be required to be deterministic, as this is required for type system soundness. This invariant is required by the compiler and cannot be broken, even with unsafe code.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Pure `const fn` under the old rules can be seen as a simple optimization opportunity for naive optimizers, as they could just reuse constant evaluation for `const fn` if the argument is known at compile time, even if the function is in a runtime context. This RFC makes such an optimization impossible. This is not seen as a problem by the author, as a more advanced optimizer (like LLVM) is able to remove these calls at compile time through means other than Rust's constant evaluation (inlining and constant folding). Also, a constant evaluation system can still evaluate executions in a runtime context, as long as it behaves exactly like runtime. The optimizer could also manually annotate functions as being truly pure by looking at the body.
+
+Secondly, with the current rules around `const fn` purity, unsafe code could choose rely on purity, e.g. by caching function return values and assuming this is not observable to clients. The author does not see this as a significant drawback, as this functionality is better served by language features that target this use case directly (like a `pure` attribute) and is therefore out of scope for the language feature of "functions evaluatable at compile time". This could break code that already relies on this, but since Rust doesn't have proper support for this, the impact should be minimal.
+
+The old rules, which say that `const fn` always has to behave the same way are already well-known in the community. Changing this will require teaching people about the new change. Since this is a very simple change, this should be easy.
+
+This is technically a breaking change. Code could rely on this behavior right now, as the [internal documentation](https://doc.rust-lang.org/1.65.0/std/intrinsics/fn.const_eval_select.html#safety) for `std::intrinsics::const_eval_select` explains. But this will not be a problem problem in practice, because all `const fn` written prior to this RFC (and also for quite some time after it until an actual way to distinguish between const eval and runtime is added) are guarateed to behave the same still. So code relying on code written prior to this RFC will still run just fine afterwards, although it did get theoretically unsound. The dependency could then of course update and introduce such difference, but this happening in practice is _highly_ unlikely and would probably require a major version bump anyways as changing behavior generally requires a major version bump.
```

Comment:
The concern is someone taking an old previously sound crate and now using it with a function that violates the guarantee. That doesn't require a major version bump anywhere.

I agree this is unlikely though, but I don't follow most of this paragraph.^^ I would just say that
- relying on this has never been officially endorsed, so technically people were relying on an unspecified aspect of the type system
- there are no known cases of code relying on this, so really this question is hypothetical and we mention this only for the sake of completeness